Mathlib/CategoryTheory/Shift/Adjunction.lean

```diff
@@ -0,0 +1,439 @@
+
+TODO: Construct a `CommShift` structure on `F` from a `CommShift` structure on `G`, using
+opposite categories.
+
+-/
+
+namespace CategoryTheory
+
+open Category
+
+namespace Adjunction
+
+variable {C D : Type*} [Category C] [Category D]
+  {F : C ⥤ D} {G : D ⥤ C} (adj : F ⊣ G) {A : Type*} [AddMonoid A] [HasShift C A] [HasShift D A]
+
+namespace CommShift
+
+variable {a b : A} (e₁ : shiftFunctor C a ⋙ F ≅ F ⋙ shiftFunctor D a)
+    (e₁' : shiftFunctor C a ⋙ F ≅ F ⋙ shiftFunctor D a)
+    (f₁ : shiftFunctor C b ⋙ F ≅ F ⋙ shiftFunctor D b)
+    (e₂ : shiftFunctor D a ⋙ G ≅ G ⋙ shiftFunctor C a)
+    (e₂' : shiftFunctor D a ⋙ G ≅ G ⋙ shiftFunctor C a)
+    (f₂ : shiftFunctor D b ⋙ G ≅ G ⋙ shiftFunctor C b)
+
+/-- Given an adjunction `adj : F ⊣ G`, `a` in `A` and commutation isomorphisms
+`e₁ : shiftFunctor C a ⋙ F ≅ F ⋙ shiftFunctor D a` and
+`e₂ : shiftFunctor D a ⋙ G ≅ G ⋙ shiftFunctor C a`, this expresses the compatibility of
+`e₁` and `e₂` with the unit of the adjunction `adj`.
+-/
+abbrev CompatibilityUnit :=
+  ∀ (X : C), (adj.unit.app X)⟦a⟧' = adj.unit.app (X⟦a⟧) ≫ G.map (e₁.hom.app X) ≫ e₂.hom.app _
+
+/-- Given an adjunction `adj : F ⊣ G`, `a` in `A` and commutation isomorphisms
+`e₁ : shiftFunctor C a ⋙ F ≅ F ⋙ shiftFunctor D a` and
+`e₂ : shiftFunctor D a ⋙ G ≅ G ⋙ shiftFunctor C a`, this expresses the compatibility of
+`e₁` and `e₂` with the counit of the adjunction `adj`.
+-/
+abbrev CompatibilityCounit :=
+  ∀ (Y : D), adj.counit.app (Y⟦a⟧) = F.map (e₂.hom.app Y) ≫ e₁.hom.app _ ≫ (adj.counit.app Y)⟦a⟧'
+
+/-- Given an adjunction `adj : F ⊣ G`, `a` in `A` and commutation isomorphisms
+`e₁ : shiftFunctor C a ⋙ F ≅ F ⋙ shiftFunctor D a` and
+`e₂ : shiftFunctor D a ⋙ G ≅ G ⋙ shiftFunctor C a`, compatibility of `e₁` and `e₂` with the
+unit of the adjunction `adj` implies compatibility with the counit of `adj`.
+-/
+lemma compatibilityCounit_of_compatibilityUnit (h : CompatibilityUnit adj e₁ e₂) :
+    CompatibilityCounit adj e₁ e₂ := by
+  intro Y
+  have eq := h (G.obj Y)
+  simp only [← cancel_mono (e₂.inv.app _ ≫ G.map (e₁.inv.app _)),
+    assoc, Iso.hom_inv_id_app_assoc, comp_id, ← Functor.map_comp,
+    Iso.hom_inv_id_app, Functor.comp_obj, Functor.map_id] at eq
+  apply (adj.homEquiv _ _).injective
+  dsimp
+  rw [adj.homEquiv_unit, adj.homEquiv_unit, G.map_comp, adj.unit_naturality_assoc, ← eq]
+  simp only [assoc, ← Functor.map_comp, Iso.inv_hom_id_app_assoc]
+  erw [← e₂.inv.naturality]
+  dsimp
+  simp only [right_triangle_components, ← Functor.map_comp_assoc, Functor.map_id, id_comp,
+    Iso.hom_inv_id_app, Functor.comp_obj]
+
+/-- Given an adjunction `adj : F ⊣ G`, `a` in `A` and commutation isomorphisms
+`e₁ : shiftFunctor C a ⋙ F ≅ F ⋙ shiftFunctor D a` and
+`e₂ : shiftFunctor D a ⋙ G ≅ G ⋙ shiftFunctor C a`, if `e₁` and `e₂` are compatible with the
+unit of the adjunction `adj`, then we get a formula for `e₂.inv` in terms of `e₁`.
+-/
+lemma CompatibilityUnit_right (h : CompatibilityUnit adj e₁ e₂) (Y : D) :
```

Comment:
Names of lemmas should start with a small letter.

Reply:
Fixed.